### DISCUSSION TITLE
Force join order

### DISCUSSION THREAD

**Discussion by D063520** [eclipse-rdf4j/rdf4j#3309] [Q&A]
Hi, given a SPARQL query like this:

select ?pub where{
  ?cancer_type wdt:P279* wd:Q12078 .
  ?pub wdt:P921 ?cancer_type.
  ?pub wdt:P31 ?class.
  ?class wdt:P279* wd:Q591041 .
 }

is there a way to force the triple pattern join order. Basically I would like to force him to the one that is given but he chooses:

```
Projection
╠══ProjectionElemList
║     ProjectionElem "pub"
╚══Join (JoinIterator)
   ├──ArbitraryLengthPath (costEstimate=14.0K, resultSizeEstimate=28.0K)
   │     Var (name=cancer_type)
   │     StatementPattern (resultSizeEstimate=22)
   │        Var (name=cancer_type)
   │        Var (name=_const_d85f4df4_uri, value=http://www.wikidata.org/prop/direct/P279, anonymous)
   │        Var (name=_const_24c6ef15_uri, value=http://www.wikidata.org/entity/Q12078, anonymous)
   │     Var (name=_const_24c6ef15_uri, value=http://www.wikidata.org/entity/Q12078, anonymous)
   └──Join (JoinIterator)
      ╠══StatementPattern (costEstimate=1.4K, resultSizeEstimate=7.9M)
      ║     Var (name=pub)
      ║     Var (name=_const_d85f6798_uri, value=http://www.wikidata.org/prop/direct/P921, anonymous)
      ║     Var (name=cancer_type)
      ╚══Join (JoinIterator)
         ├──StatementPattern (costEstimate=12.8K, resultSizeEstimate=651.4M)
         │     Var (name=pub)
         │     Var (name=_const_e5f28efe_uri, value=http://www.wikidata.org/prop/direct/P31, anonymous)
         │     Var (name=class)
         └──ArbitraryLengthPath (costEstimate=1, resultSizeEstimate=44.3M)
               Var (name=class)
               StatementPattern (resultSizeEstimate=11)
                  Var (name=class)
                  Var (name=_const_d85f4df4_uri, value=http://www.wikidata.org/prop/direct/P279, anonymous)
                  Var (name=_const_7b4d545b_uri, value=http://www.wikidata.org/entity/Q591041, anonymous)
               Var (name=_const_7b4d545b_uri, value=http://www.wikidata.org/entity/Q591041, anonymous)
```

Which I guess is far from ideal ..... because:

<img width="1024" alt="Screenshot 2021-08-18 at 23 13 08" src="https://user-images.githubusercontent.com/10883585/129973030-c47a66a5-ee5b-42c8-9175-f0bb66b0a4d3.png">

and

<img width="1024" alt="Screenshot 2021-08-18 at 23 14 40" src="https://user-images.githubusercontent.com/10883585/129973189-c5b2995b-928d-4cf9-8c26-884d826f638c.png">
 and even:

<img width="1024" alt="Screenshot 2021-08-18 at 23 15 40" src="https://user-images.githubusercontent.com/10883585/129973325-8162f658-13b0-4336-b70f-50ebd7cf273d.png">

**Comment by hmottestad** [comment]
Unfortunately not. 

Could you run the query with explanation level Executed? That way we can see how the query actually performs. https://rdf4j.org/documentation/programming/repository/#explaining-queries

If the query is sub-optimal we could look into possible rewrites to attempt to force the query optimizer to choose a better approach, or possibly consider it a bug in the query optimizer that should be fixed.

**Reply by pulquero** [reply]
Possible bug? Shouldn't the join optimizer have the two leaf-nodes swapped the other way round so the highest cardinality is always on the right arg? And given the highest cardinality tends to be on the right, shouldn't JoinIteration execute the right arg first, then for each right arg result execute the left arg, not the other way round?

**Reply by hmottestad** [reply]
I think they are ordered to account for which variables will be bound. The two bottom nodes are ordered so that they first binds based on the results from the nodes above (`pub`). If they are swapped then `class` wouldn't be bound by any previous results which means that it would need to iterate through all 44.3M statements.

**Reply by hmottestad** [reply]
It's worthwhile to mention that we don't have any functionality to calculate statistics for more than one statement pattern at a time. Being able to estimate the size of a left join on two statement patterns would be very useful.

**Reply by D063520** [reply]
"calculate statistics for more than one statement pattern at a time" -> do you think this is possible? It would be something approximate, because having an index for it sounds heavy ...

**Reply by pulquero** [reply]
Good point. How about the join execution order, left vs right first? I'm coming to the conclusion that the right recursive join structure to make life easy for the IterativeEvaluationOptimiser isn't ideal for left-as-outer join execution. I mean given this query plan, I would have thought the best strategy is right-as-outer then successively join in the left sides as you go up the tree.

(I have some custom optimizers, etc, and I'm getting in abit of a mess with what it means to be on the left vs right, wrt optimal ordering and execution, hence this thread caught my eye).

**Reply by D063520** [reply]
Not super familiar with this .... what is the difference between left and right?

**Reply by pulquero** [reply]
Currently, JoinIteration evaluates its left argument first, and then for each left bindingset, evaluate the right argument. To my mind, the right recursive join structure created by the queryoptimiser would favour the opposite evaluation order.

**Reply by D063520** [reply]
Ah, i think I understand. But then choosing between left and right would depend on the index structure, i.e. which one the index is quicker to resolve. Not sure what the current native rdf4j index prefers ....

**Reply by hmottestad** [reply]
I think by "right recursive" it means that the nested join should be placed in the right argument.

**Reply by hmottestad** [reply]
The query is optimized based on statistics provided by the store. The Native Store accounts for the different indexes when calculating the statistics. During evaluation it might be more beneficial to use a different index because more variables are bound, this is accounted for by the Native Store.

**Comment by D063520** [comment]
Hi,

thank you for this! In fact the query is running over whole Wikidata, so we do not use rdf4j as a backend, but only rdf4j as a query planner and preparer. If it is not possible we can somehow hack our planner to not do anything. Would be a nice and useful feature though ; ), no?

Salut
Dennis

**Reply by hmottestad** [reply]
How exactly do you store your data?

**Reply by D063520** [reply]
We are storing it in HDT ....

**Reply by hmottestad** [reply]
How do you expose statistics from your HDT files to the query planner?
